ldap/servers/plugins/acctpolicy/acct_plugin.c

```diff
@@ -208,75 +209,96 @@ acct_update_login_history(const char *dn, char *timestr)
     LDAPMod attribute;
     LDAPMod *list_of_mods[2];
 
-    plugin_id = get_identity();
-
-    sdn = slapi_sdn_new_normdn_byref(dn);
-    slapi_search_get_entry(&entry_pb, sdn, NULL, &e, plugin_id);
-
+    /* nothing to do if timestr is empty */
     if (!timestr) {
         return (rc);
     }
 
+    plugin_id = get_identity();
+    sdn = slapi_sdn_new_normdn_byref(dn);
+    slapi_search_get_entry(&entry_pb, sdn, NULL, &e, plugin_id);
+    slapi_sdn_free(&sdn);
+
     config_rd_lock();
     cfg = get_config();
 
+    /* does this value already exist in the entry */
+    Slapi_Value *timestr_val = slapi_value_new();
+    slapi_value_set_string(timestr_val, timestr);
+    if (slapi_entry_attr_has_syntax_value(e, cfg->login_history_attr, timestr_val)) {
+        slapi_search_get_entry_done(&entry_pb);
+        slapi_value_free(&timestr_val);
+        config_unlock();
+        return 0;
+    }
+    slapi_value_free(&timestr_val);
 
-    /* history size of zero disables login history */
-    if (cfg->login_history_size) {
-        /* get login history */
-        login_hist = slapi_entry_attr_get_charray_ext(e, cfg->login_history_attr, &num_entries);
-
-        /* first time round */
-        if (!login_hist || !num_entries) {
-            login_hist = (char **)slapi_ch_calloc(2, sizeof(char *));
-        }
-
-        /* Do we need to resize login_hist array */
+    /* now we have an entry that doesnt contain the current timestr */
+    login_hist = slapi_entry_attr_get_charray_ext(e, cfg->login_history_attr, &num_entries);
+    if (login_hist && num_entries) {
+        /* do we need to trim the array */
         if (num_entries >= cfg->login_history_size) {
+            /* free values we dont want */
             int diff = (num_entries - cfg->login_history_size);
-            /* free times we dont need */
             for (i = 0; i <= diff; i++) {
                 slapi_ch_free_string(&login_hist[i]);
             }
-            /* remap array*/
-            for (i = 0; i < (cfg->login_history_size - 1); i++) {
-                login_hist[i] = login_hist[(diff + 1) + i];
+            /* remap array if it exists */
+            for (i = 0; i < cfg->login_history_size; i++) {
+                login_hist[i] = login_hist[diff + (i + 1)];
+            }
+            if (cfg->login_history_size != 0) {
+                /* realloc and append latest value */
+                login_hist = (char **)slapi_ch_realloc((char *)login_hist, sizeof(char *) * (cfg->login_history_size + 2));
```

Comment:
realloc size is wrong ( should be login_history_size+1) and BTW I do not think that you  
should even realloc  in this case (the buffer is large enough to store the values and will be freed soon (within the pblock_destroy))